pkg/ingester/instance.go

```diff
@@ -211,6 +221,73 @@ func (i *instance) Push(ctx context.Context, req *logproto.PushRequest) error {
 	return appendErr
 }
 
+// mountGRPCError mounts an error to be returned in a GRPC call.
+//
+// The returned error is enriched with the gRPC error status and with a list of details.
+// As of now, the list of details is a list of all streams that were rate-limited. This list can be used
+//   by the distributor to fine tune streams that were limited.
+func mountGRPCError(s *stream, failedEntriesWithError []entryWithError, totalEntries int) error {
+	if len(failedEntriesWithError) == 0 {
+		return nil
+	}
+
+	lastEntryWithErr := failedEntriesWithError[len(failedEntriesWithError)-1]
+	_, ok := lastEntryWithErr.e.(*validation.ErrStreamRateLimit)
+	outOfOrder := chunkenc.IsOutOfOrderErr(lastEntryWithErr.e)
+	if !outOfOrder && !ok {
+		return lastEntryWithErr.e
+	}
+
+	var statusCode int
+	if outOfOrder {
+		statusCode = http.StatusBadRequest
+	}
+	if ok {
+		// per-stream or ingestion limited.
+		statusCode = http.StatusTooManyRequests
+	}
+
+	// Return a http status 4xx request response with all failed entries.
+	buf := bytes.Buffer{}
+	streamName := s.labelsString
+
+	limitedFailedEntries := failedEntriesWithError
+	if maxIgnore := s.cfg.MaxReturnedErrors; maxIgnore > 0 && len(limitedFailedEntries) > maxIgnore {
+		limitedFailedEntries = limitedFailedEntries[:maxIgnore]
+	}
+
+	hadPerStreamError := false
+	for _, entryWithError := range limitedFailedEntries {
+		if !hadPerStreamError {
+			if _, ok := entryWithError.e.(*validation.ErrStreamRateLimit); ok {
+				hadPerStreamError = true
+			}
+		}
+		fmt.Fprintf(&buf,
+			"entry with timestamp %s ignored, reason: '%s' for stream: %s,\n",
+			entryWithError.entry.Timestamp.String(), entryWithError.e.Error(), streamName)
+	}
+
+	fmt.Fprintf(&buf, "total ignored: %d out of %d", len(failedEntriesWithError), totalEntries)
+
+	var details []*types.Any
+
+	// Populate error details with the current stream.
+	if hadPerStreamError {
+		rls := logproto.RateLimitedStream{Labels: streamName}
+		marshalledStream, err := types.MarshalAny(&rls)
+		if err == nil {
```

Comment:
We probably want to log if `MarshalAny` has an error because it means something very strange is happening